Adding KNN and KFN CLI tests

Hsankesara opened the pull request:
This PR aims to perform complete binding and other tests in KNN and KFN. Many of the tests have been written in #1263 by @akhandait. I will add a few more test cases and make it compatible to be merged with the master.

Hsankesara commented:
Hey @rcurtin , I changed the code a bit but couldn't able to test it on my local system. I used ```make mlpack_test -j2``` command but it took too much time and eventually gave **virtual memory exhausted: Cannot allocate memory** error. I can compile *mlpack_knn* and *mlpack_kfn* directly but cannot do the same process with *knn_test* and *kfn_test*. Is there any way I can compile the test cases I want and not all?

rcurtin commented:
Hey @Hsankesara, there are a couple things you can do.  You can build without `-j2` to only use one core, which should help reduce RAM usage.  The `virtual memory exhausted` issue means your system doesn't have enough RAM to do the full build.  If you're running Linux inside of VirtualBox or something like this, maybe you can bump up the amount of RAM you have available in the virtual machine.

Another thing that you can do locally is comment out all of the files other than `knn_test.cpp` and `kfn_test.cpp` in `src/mlpack/tests/CMakeLists.txt` (a few of them will still be needed, like `mlpack_test.cpp` and some other utility files).  That should make the compilation of those parts much quicker and possibly use less RAM. :+1:

Don't check in the changes to `src/mlpack/tests/CMakeLists.txt` though---we still want the CI tests to build and run all the tests. :)

Hsankesara commented:
Hey @rcurtin, In KNN and KFN, random_basis is meant to transform the basis of current data to some random basis. But despite the fact that reference data changes when we apply the flag, the output result remains the same i.e the distance matrix and neighbor matrix. I tried to find the bug but couldn't locate it. I'd be glad if you guide me why knn and kfn are giving this type of results.

rcurtin commented:
Actually `--random_basis` doesn't change the result---it just projects both the reference set and the query set onto the same random basis, which preserves distances.  However, strangely, doing this can cause the KNN and KFN operations to be faster when using trees.  If you're trying to check in a test if it does something, I'd just take a look at the `RandomBasis()` variable on the returned `KNNModel` or `KFNModel`. :+1:

Hsankesara commented:
Thank @rcurtin, for the clarification. I took a look at the function and found that we are just transforming the basis here and indeed it shouldn't have any effect on the distance and neighbors. :+1:

Hsankesara commented:
Hey @rcurtin @akhandait, 
Please review the PR. There are three static code analysis checks warnings but the corresponding snippets are semantically correct. Please look at it once and suggest changes.

akhandait commented:
Can you take a look at why the Static code analysis checks are failing?

Hsankesara commented on Testing/Temporary/LastTest.log:
```diff
@@ -0,0 +1,3 @@
+Start testing: Mar 22 20:38 IST
+----------------------------------------------------------
+End testing: Mar 22 20:38 IST
```
Yeah! I'll delete that. Should I add it to .gitignore because we usually do not need the Testing directory in master branch?

Hsankesara commented on src/mlpack/methods/neighbor_search/kfn_main.cpp:
```diff
@@ -272,6 +272,11 @@ static void mlpackMain()
+      {
+        Log::Fatal << "Query has invalid dimensions (" << queryData.n_rows <<
+            ") whereas it should be " << kfn->Dataset().n_rows << endl;
+      }
```
No problem. I've missed that

Hsankesara commented on src/mlpack/tests/main_tests/kfn_test.cpp:
```diff
@@ -440,9 +440,9 @@ BOOST_AUTO_TEST_CASE(KFNRandomBasisTest)
   mlpackMain();
 
-  CheckMatricesNotEqual(neighbors,
+  CheckMatrices(neighbors,
```
I talked to @rcurtin before about that. He told me that changing the basis would not affect the outcomes, it speeds up the process.

akhandait replied:
Oh, I must have missed that conversation. This is okay then.

Hsankesara commented:
> Can you take a look at why the Static code analysis checks are failing?
I am trying to figure it out. But it turns out that the snippets where the warning occurred are already part of the master branch and are semantically correct.

Hsankesara commented:
Hey @akhandait, I added some suggested changes. Please check it out.

Hsankesara commented:
Hey @rcurtin and @zoq, 
Please review this PR when you've time.

akhandait commented:
Sorry for the slow response, thanks for the changes. Let's see what Ryan or Marcus have to say because I might have missed some things. Also, about adding `Testing/` to gitignore, I think it will be better if you just add it to your local gitignore. I am not sure why your code generated that while testing(an IDE thing possibly?).

Hsankesara commented:
It might be the case. I think it is generated when I ran tests locally.

Hsankesara commented:
Hey @rcurtin, 
I know you are busy. Kindly review this PR when you have time.

Hsankesara commented on src/mlpack/tests/knn_test.cpp:
```diff
@@ -8,6 +8,7 @@
  * 3-clause BSD license along with mlpack.  If not, see
  * http://www.opensource.org/licenses/BSD-3-Clause for more information.
  */
+
```
Sure, I will remove it.

Hsankesara commented on src/mlpack/tests/knn_test.cpp:
```diff
@@ -1099,7 +1101,7 @@ BOOST_AUTO_TEST_CASE(KNNModelTest)
   models[27] = KNNModel(KNNModel::TreeTypes::OCTREE, false);
 
-  for (size_t j = 0; j < 2; ++j)
+  for (size_t j = 0; j <= 2; ++j)
```
Thanks.

Hsankesara commented on src/mlpack/tests/knn_test.cpp:
```diff
@@ -249,7 +250,8 @@ BOOST_AUTO_TEST_CASE(TrainTreeTest)
   arma::Mat<size_t> tmpNeighbors(neighbors.n_rows, neighbors.n_cols);
-  for (size_t i = 0; i < distances.n_cols; ++i)
+  for (size_t i = 0; i < distances.n_cols && i < oldFromNewReferences.size();
+      ++i)
```
Yeah, it is equal to ```distances.n_cols ```. I added it just to avoid *Static Code Analysis Check*.

Hsankesara commented:
> Hey @Hsankesara, thanks for the hard work on this and sorry my review took so long. I am not actually sure if my comments are addressing @akhandait's original code or yours. :)
> 
> The tests are looking good, but there is only one that I can think of that we don't have---we should run with different leaf sizes, and then by checking the output `KNNModel`s, ensure that the tree structures themselves are different.
> 
> If you can handle those comments, I think this is ready for merge. Thanks again for working on this.

Thanks, @rcurtin, for the review. I'll make changes as soon as possible.

Hsankesara commented:
Hey @rcurtin,  my exams are starting from tomorrow. So I'll make changes after the exams(a week).

rcurtin commented:
Hey @Hsankesara, did you still have interest in working on this?  If not I'll take it over and finish it.  Let me know what you think. :+1:

Hsankesara commented:
Hey, @rcurtin, 
I apologize for leaving it for a while. I was busy in the internship. But it is now over, so I'll start working on it try to make the suggested changes in the upcoming weekend.

Hsankesara commented:
> The tests are looking good, but there is only one that I can think of that we don't have---we should run with different leaf sizes, and then by checking the output `KNNModel`s, ensure that the tree structures themselves are different.

Hey, @rcurtin, I am currently working on the test you mentioned. I am stuck and need your help. I am unable to compare two different `KNNModel` using the Boost unit test library. I tried a few methods but all of them threw multiple errors. I am wondering if there is a direct way to compare two object pointers in the Boost library or is there a better method to compare two different tree structures?

rcurtin commented:
Hey @Hsankesara, thanks for the quick response with this.  I thought about it and you are right: getting the actual tree for comparison could be quite challenging and would, I think, require writing a visitor.  I think that's overkill.  Maybe a better idea is just to check the values of `KNNModel::LeafSize()`?  What do you think?

Hsankesara commented:
Hey @rcurtin, Great idea. I can directly compare Leaf size using ```BOOOST_CHECK_EQUAL``` which would ensure that input leaf size is equal to the output one. I'll implement it as soon as possible.  Thanks for responding quickly.

Hsankesara commented:
Hey, @rcurtin, I updated the tests as we discussed and merged the master branch with my current branch. But Travis CI threw an error on three of the four build jobs. After checking the errors I found that ```ANNLayerTest``` are failing which has no relation with KNN and KFN search methods nor I changed anything on ann layer's code. As per my assessment, the error is independent of my work in this PR. Are you aware of the possible reasons for the error? 

I'd like you to review the PR and suggest possible changes if required.

rcurtin commented:
Hey @Hsankesara, yeah, the errors are not from your code.  We just merged #1953 so hopefully this should help, and I've restarted the Travis build as a result.

Hsankesara commented:
Hey @rcurtin, 
Thanks for the reply. I'll merge this branch with the master and resolve any possible conflicts.

Hsankesara commented:
Hi, @rcurtin,
I've updated the code but Travis CI is failing on QLearningTest and RadicalMainTest. They are not related to my work. I request you to review the code and check if anything needs to be changed.

rcurtin commented on src/mlpack/tests/knn_test.cpp:
```diff
@@ -1100,6 +1100,7 @@ BOOST_AUTO_TEST_CASE(KNNModelTest)
   models[26] = KNNModel(KNNModel::TreeTypes::OCTREE, true);
   models[27] = KNNModel(KNNModel::TreeTypes::OCTREE, false);
 
+
```
We can probably avoid adding these newlines.  But I can handle it during merge.

Hsankesara commented:
Thanks for approving. No issues, I can make those changes.

rcurtin commented:
Hey @Hsankesara, thanks for the contribution! :+1: I made a few changes during merge in 0019501f8, and then I realized that the tests were actually not being built in `CMakeLists.txt`, so I added them in b03ee8bd3 and 57af7320d.

I noticed that your name isn't added to the tests or to `src/mlpack/core.hpp` and `COPYRIGHT.txt`.  If you'd like to add your name, please feel free and open another PR and we can get it merged in. :)

Thanks again for your hard work with this!  Great to finally get these tests added.

Hsankesara commented:
Hey @rcurtin, thanks for merging it. I'm happy to contribute to mlpack and wish to do more in the future. 

>  realized that the tests were actually not being built in CMakeLists.txt, so I added them in b03ee8b and 57af732.

Thanks for that. I forgot about that somehow. 

> I noticed that your name isn't added to the tests or to src/mlpack/core.hpp and COPYRIGHT.txt. 

Sure, I'll open another PR for that.

rcurtin commented:
Sounds good! :+1:
